lib/kubernetes-deploy/ejson_secret_provisioner.rb

```diff
@@ -0,0 +1,168 @@
+
+      live_secrets.each do |secret|
+        secret_name = secret.metadata.name
+        next unless secret_managed?(secret)
+        next if ejson_secret_names.include?(secret_name)
+
+        KubernetesDeploy.logger.info("Pruning secret #{secret_name}")
+        @kubeclient.delete_secret(secret_name, @namespace)
+      end
+    end
+
+    def managed_secrets_exist?
+      all_secrets = @kubeclient.get_secrets(namespace: @namespace)
+      all_secrets.any? { |secret| secret_managed?(secret) }
+    end
+
+    def secret_managed?(secret)
+      secret.metadata.annotations.to_h.stringify_keys.key?(MANAGEMENT_ANNOTATION)
+    end
+
+    def encrypted_ejson
+      @encrypted_ejson ||= load_ejson_from_file
+    end
+
+    def public_key
+      encrypted_ejson["_public_key"]
+    end
+
+    def private_key
+      @private_key ||= fetch_private_key_from_secret
+    end
+
+    def validate_secret_spec(secret_name, spec)
+      errors = []
+      errors << "secret type unspecified" if spec["_type"].blank?
+      errors << "no data provided" if spec["data"].blank?
+
+      unless errors.empty?
+        raise EjsonSecretError, "Ejson incomplete for secret #{secret_name}: #{errors.join(', ')}"
+      end
+    end
+
+    def create_or_update_secret(secret_name, secret_type, data)
+      metadata = {
+        name: secret_name,
+        labels: { "name" => secret_name },
+        namespace: @namespace,
+        annotations: { MANAGEMENT_ANNOTATION => "true" }
+      }
+      secret = Kubeclient::Secret.new(type: secret_type, stringData: data, metadata: metadata)
+      if secret_exists?(secret)
+        KubernetesDeploy.logger.info("Updating secret #{secret_name}")
+        @kubeclient.update_secret(secret)
+      else
+        KubernetesDeploy.logger.info("Creating secret #{secret_name}")
+        @kubeclient.create_secret(secret)
+      end
+    rescue KubeException => e
+      raise unless e.error_code == 400
+      raise EjsonSecretError, "Data for secret #{secret_name} was invalid: #{e}"
+    end
+
+    def secret_exists?(secret)
+      @kubeclient.get_secret(secret.metadata.name, @namespace)
+      true
+    rescue KubeException => error
+      raise unless error.error_code == 404
+      false
+    end
+
+    def load_ejson_from_file
+      return {} unless File.exist?(@ejson_file)
+      JSON.parse(File.read(@ejson_file))
+    rescue JSON::ParserError => e
+      raise EjsonSecretError, "Failed to parse encrypted ejson:\n  #{e}"
+    end
+
+    def with_decrypted_ejson
+      Dir.mktmpdir("ejson_keydir") do |key_dir|
+        File.write(File.join(key_dir, public_key), private_key)
+        decrypted = decrypt_ejson(key_dir)
+        yield decrypted
+      end
+    end
+
+    def decrypt_ejson(key_dir)
+      KubernetesDeploy.logger.info("Decrypting #{EJSON_SECRETS_FILE}")
+      # ejson seems to dump both errors and output to STDOUT
+      out_err, st = Open3.capture2e("EJSON_KEYDIR=#{key_dir} ejson decrypt #{@ejson_file}")
```

Comment:
AFAICT installing [the gem](https://github.com/Shopify/ejson) just installs the Go binary... [capistrano-ejson](https://github.com/Shopify/capistrano-ejson/blob/51665d2dc7c71a7a3ad19070be993582f19b949b/lib/capistrano/tasks/ejson.cap#L14) is also shelling out to use it, as are other apps that turn up in a Github search (though I only checked a few results).